### PR TITLE
Texture atlas

### PR DESCRIPTION
**Pull request by adamhutchings**

### CODE REVIEW

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Contains all textures.
+ * @author adamhutchings
+ */
+public final class TextureAtlas {
+
+    /**
+     * The map from textures to paths.
+     */
+    private static final Map<TextureEnum, String> paths = new HashMap<>() {{
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/TextureAtlas.java:25]: use EnumMap

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            PNGDecoder decoder = null;
+
+            try {
+                decoder = new PNGDecoder(new FileInputStream(fileName));
+            } catch (IOException e) {
+                Logger.log("Could not load texture " + fileName, Severity.ERROR);
+            }
+
+            int width = decoder.getWidth();
+            int height = decoder.getHeight();
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/TextureAtlas.java:58]: decoder nullable, create throw nullpoint exception here

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Contains all textures.
+ * @author adamhutchings
+ */
+public final class TextureAtlas {
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/TextureAtlas.java:20]: add a private constructor inside the class

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java (hunk 2 above)
Comment (Shuzhengz) [reply src/org/dhsdev/flowerknight/gl/TextureAtlas.java:58]: or u can just add a `assert decoder != null;`

File: src/org/dhsdev/flowerknight/gl/Mesh.java [hunk 4]
```diff
@@ -107,18 +96,29 @@ private int getVboId(int[] arr) {
...
      * Render this mesh. This is useless right now because we don't have texture
      * code.
-     * @param textureID the OpenGL textureID to use.
+     * @param texture the OpenGL textureID to use.
      */
-    public void render(int textureID) {
+    public void render(TextureEnum texture) {
 
-        // TODO - use GLStates
+        float pos = texture.ordinal(), texMax = TextureEnum.values().length;
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/Mesh.java:103]: its good practice to declare it on separate line, easier to look at

File: src/org/dhsdev/flowerknight/gl/Mesh.java [hunk 5]
```diff
@@ -107,18 +96,29 @@ private int getVboId(int[] arr) {
...
 
-        // Activate first texture unit
-        glActiveTexture(GL_TEXTURE0);
-        GLStates.bindTexture(textureID);
+        float stride = 1.0f / texMax;
+
+        // We select a square from the large texture atlas image.
+        float[] texCoords = {
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/Mesh.java:108]: use var

**Review by Shuzhengz — CHANGES_REQUESTED** [review]
Yeah so a few small issues

**Review by Shuzhengz — DISMISSED** [review]
Cool

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * NO!
+     */
+    private TextureAtlas() {}
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/TextureAtlas.java:25]: probably wanna throw smth (illegalstateexception for example)

File: src/org/dhsdev/flowerknight/gl/TextureAtlas.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * The map from textures to paths.
+     */
+    private static final Map<TextureEnum, String> paths = new EnumMap<>(TextureEnum.class) {{
+        put(TextureEnum.LOGO, "res/logo.png");
+    }};
```
Comment (Shuzhengz) [inline src/org/dhsdev/flowerknight/gl/TextureAtlas.java:32]: use
```Java
private static final Map<TextureEnum, String> paths = new EnumMap<>(TextureEnum.class);
    
    static {
        paths.put(TextureEnum.LOGO, "res/logo.png");
    }
```

**Review by Shuzhengz — APPROVED** [review]
lgtm

**Review by Sophon96 — APPROVED** [review]
LGTM.

**Pull request merged**